File: I_AM_CHEATER/Container_With_Most_Water.js
/**
 * Time complexity in worst case, O(n^2) casue elements of height arrays are all different
 */

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  const check = (l, r) => {
    return Math.min(height[l], height[r]) * (r - l);
  };
  const set = new Set([...height].sort((a, b) => b - a));
  let answer = 0;
  for (var h of set) {
    if (answer / h > height.length) break;
    const idxs = height.flatMap((v, i) => (v >= h ? i : []));
    if (idxs.length < 2) continue;
    var ll = (() => {
      for (var i = 0; i < height.length; i++) {
        if (height[i] >= h) return i;
      }
    })();
    var rr = Math.max(...idxs);
    answer = Math.max(check(ll, rr), answer);
  }
  return answer;
};

File: I_AM_CHEATER/Container_With_Most_Water.cpp
/**
 *
 * Why the hell are you accepted
 *
 */

#include <algorithm>
#include <vector>
#include <set>

using namespace std;
class Solution
{
public:
  int maxArea(vector<int> &height)
  {
    set<int> h_set;
    for (auto h : height)
    {
      h_set.insert(h);
    }
    int answer = 0;
    for (set<int>::reverse_iterator itr = h_set.rbegin(); itr != h_set.rend(); itr++)
    {
      if (*itr == 0 || answer / (*itr) > height.size())
        break;
      int ll = findLeft(*itr, height);
      int rr = findRight(*itr, height);
      if (ll == -1 || rr == -1)
      {
        continue;
      }
      answer = max(check(ll, rr, height), answer);
    }
    return answer;
  };
  int check(int l, int r, vector<int> &height)
  {
    return min(height[l], height[r]) * (r - l);
  };
  int findLeft(int value, vector<int> &height)
  {
    for (int i = 0; i < height.size(); i++)
    {
      if (height[i] >= value)
      {
        return i;
      }
    }
    return -1;
  }
  int findRight(int value, vector<int> &height)
  {
    for (int i = height.size() - 1; i >= 0; i--)
    {
      if (height[i] >= value)
      {
        return i;
      }
    }
    return -1;
  }
};